File: json_data_extraction.py
"""Fetch JSON data from a URL or a local file and optionally persist the result."""
from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any
from urllib.parse import urlparse

import requests

DEFAULT_SOURCE = "https://raw.githubusercontent.com/emredeveloper/Database/main/db.json"


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Load JSON data from a remote URL or a local file."
    )
    parser.add_argument(
        "source",
        nargs="?",
        default=DEFAULT_SOURCE,
        help="URL or file path of the JSON resource to load.",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=Path,
        help="Optional path to save the retrieved JSON data.",
    )
    parser.add_argument(
        "--timeout",
        type=float,
        default=10.0,
        help="Timeout in seconds used for HTTP requests (default: 10).",
    )
    parser.add_argument(
        "--pretty",
        action="store_true",
        help="Pretty-print the JSON output using indentation.",
    )
    return parser.parse_args()


def is_url(value: str) -> bool:
    parsed = urlparse(value)
    return parsed.scheme in {"http", "https"}


def fetch_json_from_url(url: str, *, timeout: float) -> Any:
    try:
        response = requests.get(url, timeout=timeout)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as exc:
        raise SystemExit(f"Veri alınırken bir hata oluştu: {exc}") from exc


def load_json_from_file(path: Path) -> Any:
    try:
        with path.open("r", encoding="utf-8") as handle:
            return json.load(handle)
    except FileNotFoundError as exc:
        raise SystemExit(f"Dosya bulunamadı: {path}") from exc
    except json.JSONDecodeError as exc:
        raise SystemExit(f"JSON formatı geçersiz: {path} ({exc})") from exc


def save_json(data: Any, destination: Path, *, pretty: bool = False) -> None:
    destination.parent.mkdir(parents=True, exist_ok=True)
    with destination.open("w", encoding="utf-8") as handle:
        if pretty:
            json.dump(data, handle, ensure_ascii=False, indent=2)
        else:
            json.dump(data, handle, ensure_ascii=False)


def main() -> None:
    args = parse_args()

    source = args.source
    if is_url(source):
        data = fetch_json_from_url(source, timeout=args.timeout)
    else:
        data = load_json_from_file(Path(source))

    print("Alınan Veri:")
    if args.pretty:
        print(json.dumps(data, ensure_ascii=False, indent=2))
    else:
        print(json.dumps(data, ensure_ascii=False))

    if args.output:
        save_json(data, args.output, pretty=args.pretty)
        print(f"Veri '{args.output}' dosyasına kaydedildi.")


if __name__ == "__main__":
    main()

File: json_data_extraction.js
#!/usr/bin/env node
/**
 * Fetch JSON data from a remote URL or a local file.
 *
 * The script uses the built-in `fetch` implementation available in Node 18+.
 * When executed on older Node versions it falls back to the `node-fetch`
 * package (which must be installed separately).
 */
const fs = require("fs/promises");
const path = require("path");
const { URL } = require("url");

const DEFAULT_SOURCE = "https://raw.githubusercontent.com/emredeveloper/Database/main/db.json";

function isHttpUrl(candidate) {
  try {
    const parsed = new URL(candidate);
    return parsed.protocol === "http:" || parsed.protocol === "https:";
  } catch (error) {
    return false;
  }
}

async function ensureFetch() {
  if (typeof fetch === "function") {
    return fetch;
  }

  try {
    const { default: nodeFetch } = await import("node-fetch");
    return nodeFetch;
  } catch (error) {
    throw new Error(
      "`fetch` API kullanılabilir değil. Node 18+ sürümünü kullanın veya `npm install node-fetch` komutuyla bağımlılığı yükleyin."
    );
  }
}

async function fetchJsonFromUrl(url) {
  const fetchImpl = await ensureFetch();
  const response = await fetchImpl(url);
  if (!response.ok) {
    throw new Error(`HTTP hata! Hata kodu: ${response.status}`);
  }
  return response.json();
}

async function loadJsonFromFile(filePath) {
  const absolutePath = path.resolve(filePath);
  const contents = await fs.readFile(absolutePath, "utf-8");
  return JSON.parse(contents);
}

async function main() {
  const source = process.argv[2] ?? DEFAULT_SOURCE;

  try {
    const data = isHttpUrl(source)
      ? await fetchJsonFromUrl(source)
      : await loadJsonFromFile(source);

    console.log("Alınan Veri:");
    console.log(JSON.stringify(data, null, 2));
  } catch (error) {
    console.error("Hata:", error.message);
    process.exitCode = 1;
  }
}

main();
